StatisticsFramework/AggregatePixel.swift

```diff
@@ -0,0 +1,110 @@
+//
+//  AggregatePixel.swift
+//
+//  Copyright © 2022 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import os
+
+public actor AggregatePixel {
+
+    var lastSendDate: Date {
+        get {
+            let interval = userDefaults.double(forKey: lastSendDateKey)
+
+            // If this is new then set the last send date to now so we don't send until the next interval
+            if interval < 1 {
+                let date = Date()
+                userDefaults.set(date.timeIntervalSince1970, forKey: lastSendDateKey)
+                return date
+            }
+            return Date(timeIntervalSince1970: interval)
+        }
+
+        set {
+            let interval = newValue.timeIntervalSince1970
+            userDefaults.set(interval, forKey: lastSendDateKey)
+        }
+    }
+
+    var counter: Int {
+        get {
+            return userDefaults.integer(forKey: counterKey)
+        }
+
+        set {
+            userDefaults.set(newValue, forKey: counterKey)
+        }
+    }
+
+    let pixelName: PixelName
+    let sendInterval: TimeInterval
+    let pixel: Pixel
+    let pixelParameterName: String
+    let userDefaults: UserDefaults
+
+    var lastSendDateKey: String {
+        return "\(pixelName.rawValue).lastSendDate"
+    }
+
+    var counterKey: String {
+        return "\(pixelName.rawValue).counter"
+    }
+
+    public init(pixelName: PixelName,
+                pixelParameterName: String,
+                sendInterval: TimeInterval,
+                pixel: Pixel = Dependencies.shared.pixel,
+                userDefaults: UserDefaults = .standard) {
+
+        self.pixelName = pixelName
+        self.pixelParameterName = pixelParameterName
+        self.sendInterval = sendInterval
+        self.pixel = pixel
+        self.userDefaults = userDefaults
+    }
+
+    public func incrementAndSendIfNeeded() async {
+        counter += 1
+        await sendIfNeeded()
+    }
+
+    public func sendIfNeeded() async {
+        let lastSendInterval = lastSendDate.timeIntervalSinceNow * -1
+
+        guard lastSendInterval > sendInterval else { return }
+        if counter > 0 {
+            await fireAndReset()
```

Comment:
It's possible for this to send multiple times in quick succession if this pixel is incremented in the brief window between the send interval reaching its threshold and the pixel send completing. The counter and last send date are only reset once the pixel call has completed, so there's nothing to stop subsequent calls from calling `fireAndReset` as they have no way to tell that a pixel call is already in progress.

I had these calls milliseconds apart in Proxyman:

```
https://improving.duckduckgo.com/t/m_pageloads_with_ad_attribution_safari?count=34&test=1&test=1&macos=1
https://improving.duckduckgo.com/t/m_pageloads_with_ad_attribution_safari?count=35&test=1&test=1&macos=1
https://improving.duckduckgo.com/t/m_pageloads_with_ad_attribution_safari?count=36&test=1&test=1&macos=1
```

(I also noticed that `test=1` is being added twice to the pixel call, but I would guess this is harmless.)